src/bootstrap/lib.rs

```diff
@@ -1263,9 +1263,15 @@ impl Build {
             if !src.exists() {
                 panic!("Error: File \"{}\" not found!", src.display());
             }
-            let mut s = t!(fs::File::open(&src));
-            let mut d = t!(fs::File::create(&dst));
-            io::copy(&mut s, &mut d).expect("failed to copy");
+            let metadata = t!(src.symlink_metadata());
+            if let Err(e) = fs::copy(&src, &dst) {
+                panic!("failed to copy `{}` to `{}`: {}", src.display(),
+                       dst.display(), e)
+            }
+            t!(fs::set_permissions(&dst, metadata.permissions()));
+            let atime = FileTime::from_last_access_time(&metadata);
+            let mtime = FileTime::from_last_modification_time(&metadata);
+            t!(filetime::set_file_times(&dst, atime, mtime));
```

Comment:
So... if I'm seeing this right, the only actual change of this PR is this file.

Let me see if I get this right:

The old code copied e.g. rustc over from where it was built in stage 1 to where it is used to build the tests. This copy operation nuked all metadata and timestamps, thus causing compiletest to think rustc changed and causing a full rebuild.

The new code instead reads the metadata, copies the file and writes back the relevant old metadata. So even if there's a new file, its timestamp didn't change.

If I got everything right, please add a few comments explaining why we don't just do `fs::copy` and be done with it. r=me with that

Reply:
As far as I can tell, fs::copy doesn't copy the timestamp (just the permission bits); this code is copied from the copy method on Build (in this file) anyway, so I'd prefer to not modify it in this PR.

Reply:
To clarify, the reason we don't call *that* copy method instead is that it prefers to create a hardlink instead of copying if it can, and we don't want that here.